Add patches for (hopefully) better mod compatibility

awgv opened the pull request:
PR добавляет патчи для следующих модов:

- Vanilla Plants Expanded
- Vanilla Plants Expanded - More Plants
- Vanilla Brewing Expanded - Coffees and Teas
- Vanilla Cooking Expanded - Sushi
- Vanilla Factions Expanded - Medieval

Я проверил внутри игры, но не пропатчил:

- VFE Insectoids/Vikings/Classical
- Vanilla Genetics Expanded
- Alpha Mechs
- Vanilla Cooking Expanded
- Vanilla Cooking Expanded - Stews
- Vanilla Brewing Expanded

Потому, что стандартные унаследованные значения вроде бы работают с точки зрения баланса — насколько я смог проверить за день. Есть несколько моментов, которые я бы хотел отметить.

1. Балансировал все цифры либо относительно уже существующих в `NoMoreRiceWorld`, либо фишек модов, например, в `Vanilla Plants Expanded - More Plants` они решили, что ананасы растут долго, но более питательны, поэтому я подумал, что будет логично привязаться к тому, что они пытались добавить в игру. Скрещиваю пальцы, чтобы это было хоть сколько-нибудь сбалансированно, но должно быть всё равно лучше, чем только 100% углеводов по-умолчанию. То, что я сразу хотел поправить — добавить белки растениям, потому что иначе колония вегетарианцев будет в невыгодном положении, играя с этим модом. Я отдельно предложу PR для ванильных растений на случай, если такие изменения — это не то, что планировал автор.
2. Я не стал разбираться, как работает алкоголь, и есть ли в исходном коде что-то, что как-то на эту категорию влияет — уже устал к тому времени подбирать сбалансированные цифры для всего остального, но алкоголь даёт какое-то мизерное, 5%, кажется, повышение углеводов, что, на мой взгляд, и так сойдёт, особенно, на фоне других источников питательных веществ, поэтому этих патчей пока должно быть достаточно до первых жалоб. В `VFE - Medieval` ещё, например, бочкам, которые производят вино, не добавили гизмо для дебага, поэтому единственный способ посмотреть, как работает весь процесс — это ждать долгие внутриигровые дни. Я жду выхода пары модов и буду начинать новую партию в RimWorld, тогда посмотрю на это ещё раз в течение ближайших месяцев.
3. В `Alpha Mechs` есть робот `Pristine slurrypede`, который примерно раз в день производит `Dusgusting nutrient paste meal` из ничего (фактически энергии) — я не уверен, нужно ли это патчить каким-то образом, потому что ингридиентов он не использует, но при этом даёт по 25% всех питательных веществ. Наверное, здесь то же самое, что и с алкоголем, и поскольку ингридиентов у этой пасты нет, если что-то будет совсем несбалансированно, всегда можно добавить XML-патч конкретно для этого.
4. Мод `Alpha Animals` довольно популярен, но я никогда с ним не играл, и он добавляет очень много животных, поэтому сходу предположу, что для него тоже могут понадобиться какие-то патчи — просто для информации, если кто-то будет спрашивать. Я пока это трогать не хочу.

У меня нет твёрдого мнения на счёт цифр в этих патчах, возможно, я даже слишком сильно сфокусировался на том, чтобы найти какую-то общую точку между тем, что `NoMoreRiceWorld` делает, реальной жизнью и карикатурной симуляцией в игре, но, опять же, довольно тяжело сразу сбалансировать такое большое количество еды, поэтому, конечно же, приветствую какие-угодно изменения к этому предложению.

awgv committed:
add better mod compatibility

awgv commented:
Забыл добавить, сейчас все растения получают только кучу углеводов, может быть, имеет смысл для общей совместимости с другими модами давать растениям не 100% углеводов, а, например, не знаю, `0.4` витаминов, `0.5` углеводов и `0.1` белка? Или какие-то другие минимальные цифры. На случай, если кто-то будет играть с модом, для которого у `NoMoreRiceWorld` нет патча — чтобы не обделить такого игрока витаминами и белком. Что я могу здесь упускать?

awgv committed:
fix: patch `VCE_RawOats` only if all required mods are present

To prevent a failing patch operation.

awgv commented:
⛔ ~~Сейчас обнаружил, что некоторые патчи взрываются в зависимости от того, какой набор модов запускается — это происходит из-за перекрещивающихся зависимостей, не смерживай пока, пожалуйста, я проверю все комбинации по отдельности, чтобы в мастер не улетело. Напишу здесь, если не сегодня, то завтра.~~

Всё, вроде бы неразрешённых зависимостей больше нет. 👌

awgv committed:
fix: patch `VCE_RawLemon` only if all required mods are present

To prevent a failing patch operation.

artur-shamigulov commented:
Большое спасибо за проделанную работу, действительно впечатляющий патч. Предлагаю обсудить пару момент, затем я потрачу немного времени на тестированию и приму этот пр и зарелизжу новую версию. Не забудь добавить информацию о проделанной работе и свой профиль как контребьютор(если есть желание конечно).
1. По поводу добавления белков растениям. Моей изначальной задумкой было то, что бы заставить даже вегетарианские колонии держать животных ради получения, того что называется animal products. Но кажется что стоит сбалансировать таким образом что бы количество получаемых белков не позволяла получить баф(Возможно стоит сделать это на уровне кода, введя кап для растений)
2. Получаемый nutrition из алкоголя не окажет большого влияния на показатели
3. Предположим мех генерит еду из энергии путем конвертации ее в материю. Думаю ок, если он балансирует ингредиенты.
Опять же говоря про растения, по моему мнению, не должно быть достаточно просто выращивать всю еду и получать все бенефиты от этого. Хочется что бы была мотивация держать животных, даже если не ешь их;
Но все мои утверждения я готов обсудать.
Еще раз спасибо за предложенные патчи, я обязательно приму их после обсуждения и тестирования.

awgv commented:
> По поводу добавления белков растениям. Моей изначальной задумкой было то, что бы заставить даже вегетарианские колонии держать животных ради получения, того что называется animal products. Но кажется что стоит сбалансировать таким образом что бы количество получаемых белков не позволяла получить баф(Возможно стоит сделать это на уровне кода, введя кап для растений)

> Опять же говоря про растения, по моему мнению, не должно быть достаточно просто выращивать всю еду и получать все бенефиты от этого. Хочется что бы была мотивация держать животных, даже если не ешь их;

Тогда лучше всего будет придерживаться изначальной концепции — было бы странно ворошить авторскую идею мода. Дай мне несколько дней, пожалуйста, я погоняю игру в фоне и попробую посмотреть, какое минимальное значение протеинов на какое кол-во колонистов будет достаточно тормозить насыщение протеинами при стандартных настройках мода — может быть, дешевле с точки зрения времени будет не добавлять программный кап для растительных белков, а просто найти какую-то максимально разумную точку, что-нибудь в духе, мол, ты можешь вырастить 20 соевых бобов (или, не знаю, там, в 3 раза больше картошки) и иметь возможность избегать дебафа при изготовлении простых блюд для 3-х колонистов, а можешь завести всего несколько животных и бафнуться от души. Соответственно, чем больше колонистов, чем крупнее сама колония, и чем сложнее сами блюда, тем выгоднее иметь источник животного белка.

Иначе говоря, мне нравится идея того, что растительные протеины (но далеко не все) позволяют избегать дебафа — что это у нас, меньше 25% у колониста вроде, на самом деле, невелико значение, хочу сказать, здесь есть смысл, чтобы слишком сильно не наказывать игроков, которые по какой-то причине не имеют, возможно, временно, доступа к животным белкам или крупным фермам, все очень по-разному играют в эту игру, ну а животные источники останутся своего рода апгрейдом для белков, давая возможность без особых проблем получить баф на ~80%. Я начну с очевидных соевых бобов из мода про суши, попробую посчитать этот максимум для растительных белков, а потом размажу какую-то кривую прогрессию в меньшую сторону для всех белковых растений. Напишу здесь, как что-то будет известно, а там попробуем более конкретно переговорить, будет ли это работать, вообще.

> Не забудь добавить информацию о проделанной работе и свой профиль как контребьютор(если есть желание конечно).

Думаю, нет необходимости.

awgv commented:
Только что осознал, о чём была речь с капом растительных белков — разумеется, если каждый день будет профицит белков, рано или поздно полоска доберётся до 75%, и низкие цифры никогда не будут работать. Не знаю, о чём я думал, когда всё это писал.

Если говорить о реализации капа растительных белков (если имелся в виду предел в буквальном смысле), было бы здорово скрыть кап и позволять полоске заполняться до 100%, не давая бафа, чтобы игроки не недоумевали, почему белок не поднимается выше 70%, и не выделяться на фоне стандартного интерфейса. Единственное, меня смущает, что кап, как бы он ни был реализован, это в любом случае что-то скрытое, может быть, это решается подсказкой в описании `NeedDef` или даже уведомлением письмом — имеет смысл об этом подумать.

Я планирую отправить на днях в оба PR обнуление белков, чтобы их можно было смержить в текущем виде в ближайшие недели, а если ты когда-нибудь решишь добавить логику конкретно для белков, я лучше открою ещё один.

awgv committed:
reset proteins back to zero for all plants

Also, merge most patches into one operation, because the minute difference in vitamins/carbs probably won’t matter in the long run—keeping it simple as intended.
